From: Yaroslav Halchenko <debian@onerussian.com>
Subject: API to control LIBSVM verbosity without patching

  ATM scikits.learn uses preprocessor's ifdefs to remove verbose
  output -- it would be better to provide an ability for a user to get
  control over the verbosity within Python interface.

  By default -- no verbose output should be printed

--- a/scikits/learn/svm/src/libsvm/libsvm_helper.c
+++ b/scikits/learn/svm/src/libsvm/libsvm_helper.c
@@ -669,5 +669,33 @@ int free_param(struct svm_parameter *par
     free(param);
     return 0;
 }
-       
-                            
+
+/* rely on built-in facility to control verbose output
+ * in the versions of libsvm >= 2.89
+ */
+#if LIBSVM_VERSION && LIBSVM_VERSION >= 289
+
+/* borrowed from original libsvm code */
+static void print_null(const char *s) {}
+
+static void print_string_stdout(const char *s)
+{
+	fputs(s,stdout);
+	fflush(stdout);
+}
+
+/* provide convenience wrapper */
+void set_verbosity(int verbosity_flag){
+	if (verbosity_flag)
+# if LIBSVM_VERSION < 291
+		svm_print_string = &print_string_stdout;
+	else
+		svm_print_string = &print_null;
+# else
+		svm_set_print_string_function(&print_string_stdout);
+	else
+		svm_set_print_string_function(&print_null);
+# endif
+}
+#endif
+
--- a/scikits/learn/svm/src/libsvm/_libsvm.pyx
+++ b/scikits/learn/svm/src/libsvm/_libsvm.pyx
@@ -93,7 +93,7 @@ cdef extern from "libsvm_helper.c":
     int  free_model     (svm_model *)
     int  free_model_SV  (svm_model *)
     int  free_param     (svm_parameter *)
-
+    void set_verbosity(int)
 
 ################################################################################
 # Wrapper functions
@@ -579,3 +579,8 @@ def predict_margin_from_model_wrap(np.nd
     free_param(param)
     return dec_values
 
+def set_verbosity_wrap(int verbosity):
+    """
+    Control verbosity of libsvm library
+    """
+    set_verbosity(verbosity)
--- a/scikits/learn/svm/base.py
+++ b/scikits/learn/svm/base.py
@@ -352,3 +352,5 @@ class BaseLibLinear(BaseEstimator):
         """
         return int  (self.fit_intercept) - .5
 
+
+_libsvm.set_verbosity_wrap(0)
